lib/handler/range.c

```diff
@@ -0,0 +1,124 @@
+#include <assert.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "h2o.h"
+
+
+#define CHECK_EOF()                             \
+    if (buf == buf_end) {                       \
+        *ret = -2;                              \
+        return NULL;                            \
+    }
+
+#define EXPECT_CHAR(ch)                         \
+    CHECK_EOF();                                \
+    if (*buf++ != ch) {                         \
+        *ret = -1;                              \
+        return NULL;                            \
+    }
+
+#define CHECK_OVERFLOW(range)                   \
+    if (range == SIZE_MAX) {                    \
+        *ret = -1;                              \
+        return NULL;                            \
+    }
+
+size_t *process_range(h2o_mem_pool_t *pool, h2o_iovec_t *range_value, size_t file_size, size_t* ret)
+{
+    size_t range_start = -1, range_count = 0;
+    char *buf = range_value->base, *buf_end = buf + range_value->len;
+    int good_range = 1, ows_skipped = 0;
+    H2O_VECTOR(size_t) ranges = {};
+    
+    if (range_value->len < 6 || memcmp(buf, "bytes=", 6) != 0) {
+        *ret = -1;
+        return NULL;
+    }
+
+    buf += 6;
+    
+    /* most range requests contain only one range */
+    do {
+        while (1) {
+            if (*buf != ',') {
+                if (ows_skipped) {
+                    *ret = -1;
+                    return NULL;
+                }
+                break;
+            }
+            ows_skipped = 0;
+            buf ++;
+            while (H2O_UNLIKELY(*buf == ' ') || H2O_UNLIKELY(*buf == '\t')) {
+                buf ++;
+                CHECK_EOF();
+            }
+        }
+        if (H2O_UNLIKELY(buf == buf_end)) break;
+        range_start = -1; range_count = 0;
```

Comment:
Please use `SIZE_MAX` instead of `-1` when assigning ~0 to `size_t`.

I would also suggest removing variable `good_range`, ad using `range_start != SIZE_MAX` for determining if the parser has succeeded in extracting a valid range.